Webservice product az (#1703)
PrestaShop/docs: webservice/tutorials/create-product-az.md

---

> title: Create aPproduct from A to Z with the Webservices

Author: ```suggestion
title: Create a product from A to Z with the Webservices
```

> title: Create a product from A to Z with the Webservices

Reviewer: ```suggestion
title: Create a product from start to finish with Webservices
```

> menuTitle: Product from A to Z

Reviewer: ```suggestion
menuTitle: Creating a product - complete guide
```

weight: 6
---

> # Create a product from A to Z with the Webservices

Reviewer: ```suggestion
# Create a product from start to finish with Webservices
```

In this guide, we will show you how to handle the creation and manipulation of a `Product`, and its dependencies, from the Webservices. 

> We will first create a `Supplier` and a `Manufacturer`. Secondly, we will create a `Category` to put our `Product` in. 

Reviewer: ```suggestion
This guide will demonstrate how to create and manipulate a `Product` and its dependencies from the Webservices.

Firstly, we will generate a "Supplier" and a "Manufacturer". Secondly, we will establish a "Category" to organize our "Product".
```

> From this point, we will be able to add the `Product`. 

Reviewer: ```suggestion
```

> When the `Product` will be added, we will add some `Stocks`, upload an `Image`, and manage some multi-language content. 

Reviewer: ```suggestion
Once the `Product` is added, we'll include some `Stocks`, upload an `Image`, and handle multi-language content.
```

> We will illustrate those examples [with a Postman Collection]({{<relref "/8/webservice/tutorials/testing-webservice-postman">}}), that you can download and run against your own's PrestaShop Webservices. 

Reviewer: ```suggestion
We will illustrate those examples [with a Postman Collection]({{<relref "/8/webservice/tutorials/testing-webservice-postman">}}), that you can download and run against your own PrestaShop Webservices. 
```

> [Download the Postman Collection](https://github.com/PrestaShop/webservice-postman-examples/blob/main/demo_product_az_collection.json)

Author: ```suggestion
[Download the Postman Collection](https://github.com/PrestaShop/webservice-postman-examples/blob/main/demo_product_az.json)
```

## Create a Supplier and Manufacturer

> Altough optional, having `Manufacturer` and `Supplier` will help managing your stock and business. 

Reviewer: ```suggestion
Although optional, having both `Manufacturer` and `Supplier` information will assist in managing your stock and business efficiently.
```

You have to send a request for each `Manufacturer` and `Supplier` to the Webservices before creating your `Product`. 

### Create a Manufacturer

> Send a `POST` request to `/api/manufacturers` with this body: 

Reviewer: ```suggestion
Send a `POST` request to `/api/manufacturers` with the following body: 
```

        <name>
            <![CDATA[ACMEManufacturer]]>
        </name>
        <date_add>
            <![CDATA[2023-08-07 09:31:06]]>
        </date_add>
        <date_upd>
            <![CDATA[2023-08-07 09:31:06]]>
        </date_upd>
        <description>
            <language id="1">
                <![CDATA[]]>
            </language>
        </description>
        <short_description>
            <language id="1">
                <![CDATA[]]>
            </language>
        </short_description>
        <meta_title>
            <language id="1">
                <![CDATA[]]>
            </language>
        </meta_title>
        <meta_description>
            <language id="1">
                <![CDATA[]]>
            </language>
        </meta_description>
        <meta_keywords>
            <language id="1">
                <![CDATA[]]>
            </language>
        </meta_keywords>
        <associations>
            <addresses nodeType="address" api="addresses"/>
        </associations>
    </manufacturer>
</prestashop>
```

{{% notice note %}}
Related request in the Postman collection: 

1 - Create manufacturer
{{% /notice %}}

### Create a Supplier

> Send a `POST` request to `/api/suppliers` with this body: 

Reviewer: ```suggestion
Send a `POST` request to `/api/suppliers` with the following body: 
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
    <supplier>
        <name>ACMESupplier</name>
    </supplier>
</prestashop>
```

Retrieve the newly created supplier `ID` from the Webservices's response.

{{% notice note %}}
Related request in the Postman collection: 

2 - Create supplier
{{% /notice %}}

## Create a Category for our Product

> We will then create a `Category` to put our `Product` in. 

Reviewer: ```suggestion
Let's create a designated `Category` for our `Product`.
```

> Send a `POST` request to `/api/categories` with this body:

Reviewer: ```suggestion
Send a `POST` request to `/api/categories` with the following body:
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
<category>
    <name>
        <language id="1"><![CDATA[Category demo]]></language>
    </name>
    <link_rewrite>
        <language id="1"><![CDATA[category-demo]]></language>
    </link_rewrite>
    <description>
        <language id="1"><![CDATA[my awesome category description]]></language>
    </description>
    <active>1</active>
    <id_parent>1</id_parent>
</category>
</prestashop>
```

> 
> Retrieve the newly created category `ID` from the Webservices's response.

Reviewer: ```suggestion
Ensure the provided `Language` ID corresponds to your `Language` ID in your back office.

Retrieve the newly created category `ID` from the Webservices's response.
```

{{% notice note %}}
Related request in the Postman collection: 

3 - Create category
{{% /notice %}}

## Create the Product

> Now that we have all minimum `Product` dependencies, we can create a minimum `Product`.

Reviewer: ```suggestion
Now that the required dependencies are in place, we can move forward with crafting the minimum viable product.
```

> Send a `POST` request to `/api/products` with this body: 

Reviewer: ```suggestion
Send a `POST` request to `/api/products` with the following body: 
```

    <price><![CDATA[123.45]]></price>
    <unit_price><![CDATA[123.45]]></unit_price>
    <active><![CDATA[1]]></active>
    <meta_description>
        <language id="1"><![CDATA[Description]]></language>
    </meta_description>
    <meta_keywords>
        <language id="1"><![CDATA[Keywords]]></language>
    </meta_keywords>
    <meta_title>
        <language id="1"><![CDATA[My Title for SEO]]></language>
    </meta_title>
    <link_rewrite>
        <language id="1"><![CDATA[awesome-product]]></language>
    </link_rewrite>
    <name>
        <language id="1"><![CDATA[My awesome Product]]></language>
    </name>
    <description>
        <language id="1"><![CDATA[Description]]></language>
    </description>
    <description_short>
        <language id="1"><![CDATA[Short description]]></language>
    </description_short>
    <associations>
        <categories>
            <category>
                <id><![CDATA[1]]></id>
            </category>
        </categories>
    </associations>
</product>
</prestashop>
```

Replace `id_manufacturer`, `id_supplier`, `id_category_default`, `associations > categories > category > id` with the retrieved IDs from the previously created entities.

> This will create a `Product`, of type: `ProductType::TYPE_STANDARD`, associated to the shop with `id:1`.

Reviewer: ```suggestion
This will create a `Product`, of type: `ProductType::TYPE_STANDARD`, associated with the `Shop` with `id:1`.
```

{{% notice note %}}
Related request in the Postman collection: 

4 - Create product
{{% /notice %}}

> ## Add some Stock to Product

Reviewer: ```suggestion
## Product Stock management
```

> When a `Product` entity is created, a matching `Stock_Available` entity is created. 

Reviewer: ```suggestion
Whenever a `Product` entity is formed, a corresponding `Stock_Available` record is also generated and inserted to the database.
```

{{% notice note %}}
Related request in the Postman collection: 

5 - Get stock available
{{% /notice %}}

```xml
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
    <stock_availables>
        <stock_available>
            <id>
                <![CDATA[67]]>
            </id>
            <id_product xlink:href="http://localhost:8080/api/products/27">
                <![CDATA[27]]>
            </id_product>
            <id_product_attribute>
                <![CDATA[0]]>
            </id_product_attribute>
            <id_shop xlink:href="http://localhost:8080/api/shops/1">
                <![CDATA[1]]>
            </id_shop>
            <id_shop_group>
                <![CDATA[0]]>
            </id_shop_group>
            <quantity>
                <![CDATA[0]]>
            </quantity>
            <depends_on_stock>
                <![CDATA[0]]>
            </depends_on_stock>
            <out_of_stock>
                <![CDATA[2]]>
            </out_of_stock>
            <location>
                <![CDATA[]]>
            </location>
        </stock_available>
    </stock_availables>
</prestashop>
```

> Then, update the available quantities by sending a `PATCH` request to `/api/stock_availables/{{id}}` (with `{{id}}` the previously retrieved stock_available entity id). 

Reviewer: ```suggestion
You can update the available quantities by sending a `PATCH` request to `/api/stock_availables/{{id}}` (with `{{id}}` the previously retrieved stock_available entity id). 
```

In this example, the returned stock_available entity id was: `67`.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
    <stock_available>
        <id><![CDATA[67]]></id>
        <quantity><![CDATA[10]]></quantity>
    </stock_available>
</prestashop>
```

{{% notice note %}}
Related request in the Postman collection: 

5 - Update stock available
{{% /notice %}}

### Complete update of the Stock_available entity

You can also update the whole `Stock_available` entity by changing the `PATCH` request to a `PUT` request, and provide all the fields. 
 
{{% notice note %}}
Related request in the Postman collection: 

5 - Update stock available - complete
{{% /notice %}}

{{% notice info %}}

> Depending on your stock management strategy (`Share available quantities for sale` between shops, in `back office > Advanced Parameters > Multistore > Edit shop group`), you need to provide an `id_shop` (per shop stock) or an `id_shop_group` (shared shop stock) to the request.

Reviewer: ```suggestion
Depending on your stock management strategy (`Share available quantities for sale` between shops, in the `Back office > Advanced Parameters > Multistore > Edit shop group`), you need to provide an `id_shop` (per shop stock) or an `id_shop_group` (shared shop stock) to the request.
```

{{% /notice %}}

> ## Upload an Image on the Product

Reviewer: ```suggestion
## Product images management
```

> To upload an Image on your Product, send a `POST` request to `/api/images/products/{{product_id}}` (with `{{id}}` the newly created Product entity id). 

Reviewer: ```suggestion
To upload an `Image` to your `Product`, send a `POST` request to `/api/images/products/{{product_id}}` (with `{{id}}` of the newly created `Product` entity). 
```

Add an `image` form-data input in the body of the request, with your image content attached. 

{{% notice note %}}
Related request in the Postman collection: 

6 - Upload image
{{% /notice %}}

## Manage multi-lang content

> Considering you have 2 languages set in your PrestaShop instance, let's say: 

Reviewer: ```suggestion
If you have two languages set up in your PrestaShop instance, for example:
```

            <language id="2"><![CDATA[Mots clés FR]]></language>
        </meta_keywords>
        <meta_title>    
            <language id="1"><![CDATA[English Title]]></language>
            <language id="2"><![CDATA[Titre en Francais]]></language>
        </meta_title>
        <link_rewrite>
            <language id="1"><![CDATA[english-title]]></language>
            <language id="2"><![CDATA[titre-francais]]></language>
        </link_rewrite>
        <name>
            <language id="1"><![CDATA[Product name]]></language>
            <language id="2"><![CDATA[Nom du produit]]></language>
        </name>
        <description>
            <language id="1"><![CDATA[English Description]]></language>
            <language id="2"><![CDATA[Description en français]]></language>
        </description>
        <description_short>
            <language id="1"><![CDATA[Short English description]]></language>
            <language id="2"><![CDATA[Description courte en français]]></language>
        </description_short>
    </product>
</prestashop>
```

{{% notice note %}}
Related request in the Postman collection: 

7 - Update multi-lang fields
{{% /notice %}}

> ## Create a ProductType::TYPE_COMBINATIONS Product

Reviewer: ```suggestion
## Create a Product with Combinations - ProductType::TYPE_COMBINATIONS
```

> To create a Combination Product Type, you need to create your `Product` and then your `Combinations`. 

Reviewer: ```suggestion
To create a Product with Combinations, you must make your `Product` and then add your `Combinations`. 
```

### Create the Product

> Change in the xml body of the [Product creation request]({{< relref "#create-the-product" >}}), on the `<product_type>` node, the value from `standard` to `combinations`.

Reviewer: ```suggestion
Change XML body of the [Product creation request]({{< relref "#create-the-product" >}}). Modify `<product_type>` node, and replace `standard` with `combinations`.
```

> Send a `POST` request to `/api/products` with this body: 

Reviewer: ```suggestion
Send a `POST` request to `/api/products` with the following body: 
```

        <meta_description>
            <language id="1"><![CDATA[Description]]></language>
        </meta_description>
        <meta_keywords>
            <language id="1"><![CDATA[Keywords]]></language>
        </meta_keywords>
        <meta_title>
            <language id="1"><![CDATA[My Title for SEO]]></language>
        </meta_title>
        <link_rewrite>
            <language id="1"><![CDATA[awesome-product]]></language>
        </link_rewrite>
        <name>
            <language id="1"><![CDATA[My awesome Product with combinations]]></language>
        </name>
        <description>
            <language id="1"><![CDATA[Description]]></language>
        </description>
        <description_short>
            <language id="1"><![CDATA[Short description]]></language>
        </description_short>
        <associations>
            <categories>
                <category>
                    <id><![CDATA[1]]></id>
                </category>
            </categories>
        </associations>
    </product>
</prestashop>
```

{{% notice note %}}
Related request in the Postman collection: 

8 - Create Combination Product
{{% /notice %}}

### Create the Combinations

> Send two `POST` requests to `/api/combinations` to create your combinations, with `{{id_product}}` your previously created product id. 

Reviewer: ```suggestion
Send two `POST` requests to `/api/combinations` to create your combinations. Use `{{id_product}}` of your previously created `Product`. 
```

Why do we need to send it twice? Also, it seems there's no information about the combination, like... which group of the attribute is thta? what's the value?

Reviewer: cc @thomasnares

Author: yes i still need to change this paragraph and explain about attributes 👍

Reviewer: ok, I converted your PR to draft :D let me know once it's ready
